File: kotlin/src/mal/step3_env.kt
fun READ(s: String) = read_str(s)

// Create a new function eval_ast which takes ast (mal data type) and
// an associative structure (the environment from above). eval_ast
// switches on the type of ast as follows:
// 
//  * symbol: lookup the symbol in the environment structure and return the value or raise an error if no value is found
//  * list: return a new list that is the result of calling EVAL on each of the members of the list
//  * otherwise just return the original ast value

fun eval_ast(ast: MalType, env: Env, depth: Int) : MalType {
//    print("eval_ast: ".repeat(depth))
//    println(PRINT(ast))
    return when(ast) {
        is MalList   -> MalList(ast.atoms.map { EVAL(it, env, depth) }.toList())
        is MalSymbol -> env.get(ast)
        else -> ast
    }
}

// Modify EVAL to check if the first parameter ast is a list.
// * ast is not a list: then return the result of calling eval_ast on it.
// * ast is a empty list: return ast unchanged.
// * ast is a list: call eval_ast to get a new evaluated list. Take the first
//   item of the evaluated list and call it as function using the rest of the
//   evaluated list as its arguments.

/*
Modify the apply section of EVAL to switch on the first element of the
list:

    symbol "def!": call the set method of the current environment
(second parameter of EVAL called env) using the unevaluated first
parameter (second list element) as the symbol key and the evaluated
second parameter as the value.

    symbol "let*": create a new environment using the current
environment as the outer value and then use the first parameter as a
list of new bindings in the "let*" environment. Take the second
element of the binding list, call EVAL using the new "let*"
environment as the evaluation environment, then call set on the "let*"
environment using the first binding list element as the key and the
evaluated second element as the value. This is repeated for each
odd/even pair in the binding list. Note in particular, the bindings
earlier in the list can be referred to by later bindings. Finally, the
second parameter (third element) of the original let* form is
evaluated using the new "let*" environment and the result is returned
as the result of the let* (the new let environment is discarded upon
completion).

    otherwise: call eval_ast on the list and apply the first element to the rest as before.

*/

fun make_env(pairs: MalList, outer_env: Env, depth: Int) : Env {
    val new_env = Env(outer_env)
    for (idx in pairs.atoms.indices step 2) {
        val k = pairs.atoms[idx] as MalSymbol
        val v = pairs.atoms[idx + 1]
        new_env.set(k, EVAL(v, new_env, depth + 1))
    }
    return new_env
}

var eval_count = 0
fun EVAL(ast: MalType, env: Env, depth: Int) : MalType {
//    print("EVAL____: ".repeat(depth))
//    println(PRINT(ast))

    eval_count += 1
    if (depth > 200 || eval_count > 500) {
        throw Exception("Recursion fail :(")
    }

    if (ast !is MalList) {
        return eval_ast(ast, env, depth)
    }
    else {
        if (ast.atoms.isEmpty()) {
            return ast
        }
        else {
            val op   = ast.head()
            val args = ast.tail()
            if(op is MalSymbol) {
                when(op.sym) {
                    "def!" -> {
                        val v = EVAL(args[1], env, depth + 1)
                        return env.set((args[0] as MalSymbol), v)
                    }
                    "let*" -> return EVAL(args[1], make_env((args[0] as MalList), env, depth), depth + 1)
                }
            }
            val l = eval_ast(ast, env, depth + 1)
            val f = ((l as MalList).head() as MalFunc)
            return f(l.tail())
        }
    }
}

fun PRINT(v: MalType) = pr_str(v)

val repl_env = Env(null).apply {
    set(MalSymbol("+"), MalFunc({ a, b -> MalNumber((a as MalNumber).num + (b as MalNumber).num) }))
    set(MalSymbol("-"), MalFunc({ a, b -> MalNumber((a as MalNumber).num - (b as MalNumber).num) }))
    set(MalSymbol("*"), MalFunc({ a, b -> MalNumber((a as MalNumber).num * (b as MalNumber).num) }))
    set(MalSymbol("/"), MalFunc({ a, b -> MalNumber((a as MalNumber).num / (b as MalNumber).num) }))
}

fun rep(s: String) {
    println(PRINT(EVAL(READ(s), repl_env, 0)))
}

fun main(args: Array<String>) {
    while(true) {
        print("user> ")

        try {
            readLine()?.let { rep(it) }
        }
        catch(e: Exception) {
            println("Oh dear:" + e.toString())
        }
    }
}

File: impls/ts/step3_env.ts
import { MalNumber, MalList, MalMap, MalType, mal, MalFunc, MalSymbol } from './types.ts'
import { read_str } from './reader.ts'
import { pr_str } from './printer.ts'
import Env from './env.ts';

function eval_ast(ast: MalType, env: Env): MalType {
    switch(ast.type) {
        case 'symbol': {
            return env.get(ast)
        }
        case 'list':
        case 'vector': {
            const valGen = mal[ast.type]
            const seed = valGen([])
                return ast.values.reduce(
                    (a: typeof seed, b: MalType) => valGen(a.values.concat([EVAL(b, env)])),
                    seed
                )
            }
        case 'map': {
            return Array.from(ast.values.entries()).reduce(
                (acc: MalMap, [k, v]) => mal.map(acc.values.set(k, EVAL(v, env))),
                mal.map(new Map())
            )
        }
        default:
            return ast
    }
}

function READ(s: string): MalType {
    return read_str(s)
}
function EVAL(v: MalType, env: Env): MalType {
    if(v.type != 'list') {
        return eval_ast(v, env)
    }
    else if(v.values.length === 0) {
        return v
    }
    else {
        const head = v.values[0]
        const tail = v.values.slice(1)

        if(head.type === 'symbol') {
            switch(head.value) {
                case 'def!': {
                    const v = EVAL(tail[1], env)
                    return env.set((tail[0] as MalSymbol), v)
                }
                case 'let*': {
                    return EVAL(tail[1], makeEnv((tail[0] as MalList), env))
                }
            }
        }

        const callSite = eval_ast(v, env) as MalList
        const f        = callSite.values[0] as MalFunc
        const args     = callSite.values.slice(1)

        return f.value.apply(null, args)
    }
}
function PRINT(v: MalType): string {
    return pr_str(v)
}
function rep(s: string): string {
    return PRINT(EVAL(READ(s), env))
}

function makeEnv(pairs: MalList, outerEnv: Env): Env {
    const newEnv = new Env(outerEnv)
    for(let idx = 0; idx < pairs.values.length; idx += 2) {
        const k = pairs.values[idx] as MalSymbol
        const v = pairs.values[idx + 1]
        newEnv.set(k, EVAL(v, newEnv))
    }
    return newEnv
}

function malCalc(f: (x: number, y: number) => number): MalFunc {
    return mal.func((a: MalType, b: MalType) => {
        const x = a as MalNumber
        const y = b as MalNumber
        return mal.num(f(x.value, y.value))
    })
}

const env = makeEnv(
    mal.list([
        mal.symbol('+'), malCalc((a,b) => a + b),
        mal.symbol('-'), malCalc((a,b) => a - b),
        mal.symbol('*'), malCalc((a,b) => a * b),
        mal.symbol('/'), malCalc((a,b) => a / b),
    ]),
    new Env()
)

let line = prompt('user>')
while (line !== null) {
    try {
        console.log(rep(line));
    }
    catch(e) {
        console.log('Error: ', e)
    }
    line = prompt('user>')
}
